local-dev-environment/search_client.py

```diff
@@ -0,0 +1,174 @@
+SAFE_SEARCH_TERM = str(SEARCH_TERM).replace("/", "_").replace(" ", "_")
+TIMESTAMP_STR = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+OUTPUT_PATH = OUTPUT_DIRECTORY / f"{TIMESTAMP_STR}_{SAFE_SEARCH_TERM}_search_results.xlsx"
+
+# --- 3. DEFINE K-NN SEARCH QUERY ---
+
+
+def create_hybrid_query(query_text: str, query_vector: list[float], k: int = 5) -> dict:
+    """Create a hybrid search query combining keyword and semantic vector search."""
+    if not FUZZY:
+        return {
+            "size": k,
+            "_source": ["document_id", "page_number", "chunk_text", "case_ref"],
+            "query": {
+                "bool": {
+                    "should": [
+                        {"match": {"chunk_text": {"query": query_text, "boost": KEYWORD_BOOST}}},
+                        {"knn": {"embedding": {"vector": query_vector, "k": k, "boost": SEMANTIC_BOOST}}},
+                    ]
+                }
+            },
+        }
+    else:
+        return {
+            "size": k,
+            "_source": ["document_id", "page_number", "chunk_text", "case_ref"],
+            "query": {
+                "bool": {
+                    "should": [
+                        {
+                            "fuzzy": {
+                                "chunk_text": {
+                                    "value": query_text,
+                                    "fuzziness": FUZZINESS,
+                                    "max_expansions": MAX_EXPANSIONS,
+                                    "boost": FUZZY_BOOST,
+                                }
+                            }
+                        },
+                        {"knn": {"embedding": {"vector": query_vector, "k": k, "boost": SEMANTIC_BOOST}}},
+                    ]
+                }
+            },
+        }
+
+
+# --- 4. Execute results and write to Excel ---
+def local_search_client() -> list[dict]:
+    """Execute a hybrid search on the local OpenSearch instance and return the hits."""
+    logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger("embedding_search_client")
+    try:
+        embedding_generator = EmbeddingGenerator(settings.BEDROCK_EMBEDDING_MODEL_ID)
+        embedding = embedding_generator.generate_embedding(SEARCH_TERM)
+        logger.info(f"Generated embedding for search term: '{SEARCH_TERM}'")
+
+        client = OpenSearch(
+            hosts=[settings.OPENSEARCH_PROXY_URL],
+            http_auth=(USER, PASSWORD),
+            use_ssl=False,
+            verify_certs=False,
+            ssl_assert_hostname=False,
+        )
+
+        search_query = create_hybrid_query(SEARCH_TERM, embedding, k=K_QUERIES)
+        logger.info(f"Performing hybrid search for {K_QUERIES} neighbors in '{CHUNK_INDEX_NAME}'...")
+        response = client.search(index=CHUNK_INDEX_NAME, body=search_query)
+
+        hits = response["hits"]["hits"]
+
+        # print(f"hits: {hits}")
+        return hits
+
+    except ConnectionError as ce:
```

Comment:
Good use of exception handling, but you might want to catch opensearchpy.exceptions.ConnectionError specifically, rather than the built-in ConnectionError. Have a look at that see if it would be suitable.